packages/ra-core/src/form/FormDataConsumer.tsx

```diff
@@ -43,22 +43,20 @@ import warning from '../util/warning';
  * );
  */
 const FormDataConsumer = (props: ConnectedProps) => {
-    const formData = useWatch();
+    const { getValues } = useFormContext();
 
-    return <FormDataConsumerView formData={formData} {...props} />;
+    return <FormDataConsumerView formData={getValues()} {...props} />;
```

Comment:
I'm afraid this is not strictly equivalent, as `formData` is actually (I believe, kinda like) a state, that will trigger re-render when it changes, but `getValues()` is actually a callback, that only works when we call it.

The following `PostEdit` (based on [this doc example](https://marmelab.com/react-admin/Inputs.html#linking-two-inputs)) no longer works with your change:

```tsx
import * as React from 'react';
import {
    ArrayInput,
    Edit,
    FormDataConsumer,
    SelectInput,
    SimpleForm,
    SimpleFormIterator,
    TextInput,
} from 'react-admin'; // eslint-disable-line import/no-unresolved

const PostEdit = () => (
    <Edit>
        <SimpleForm>
            <ArrayInput source="authors">
                <SimpleFormIterator>
                    <TextInput source="name" />
                    <FormDataConsumer>
                        {({
                            formData, // The whole form data
                            scopedFormData, // The data for this item of the ArrayInput
                            getSource, // A function to get the valid source inside an ArrayInput
                            ...rest
                        }) =>
                            scopedFormData && scopedFormData.name ? (
                                <SelectInput
                                    source={getSource('role')} // Will translate to "authors[0].role"
                                    choices={[
                                        { id: 1, name: 'Head Writer' },
                                        { id: 2, name: 'Co-Writer' },
                                    ]}
                                    {...rest}
                                />
                            ) : null
                        }
                    </FormDataConsumer>
                </SimpleFormIterator>
            </ArrayInput>
        </SimpleForm>
    </Edit>
);

export default PostEdit;

```

Reply:
Great review!